demos/GTK/Gtk4FunctionPlotDemo/MainWindow.cs

```diff
@@ -0,0 +1,121 @@
+    private static readonly double     s_funcMin;
+    private static readonly double     s_funcMax;
+
+    private ImageSurface? _functionImageSurface;
+    private MousePosition _mousePosition;
+    private bool          _mouseIsInDrawingArea;
+    //-------------------------------------------------------------------------
+    public MainWindow(Application app)
+    {
+        this.Application = app;
+        this.Title       = "Peaks function plot";
+        this.Resizable   = false;
+
+        this.Child = this.CreateDrawingArea();
+    }
+    //-------------------------------------------------------------------------
+    private DrawingArea CreateDrawingArea()
+    {
+        DrawingArea drawingArea   = DrawingArea.New();
+        drawingArea.ContentWidth  = Size;
+        drawingArea.ContentHeight = Size;
+
+        drawingArea.MarginStart   = 10;
+        drawingArea.MarginTop     = 10;
+        drawingArea.MarginEnd     = 10;
+        drawingArea.MarginBottom  = 10;
+
+        drawingArea.SetDrawFunc(this.Draw);
+
+        EventControllerMotion mouseMove = EventControllerMotion.New();
+        drawingArea.AddController(mouseMove);
+
+        mouseMove.OnMotion += (EventControllerMotion sender, EventControllerMotion.MotionSignalArgs args) =>
+        {
+            _mousePosition.X = args.X;
+            _mousePosition.Y = args.Y;
+
+            (sender.Widget as DrawingArea)?.QueueDraw();
+        };
+
+        mouseMove.OnEnter += (EventControllerMotion sender, EventControllerMotion.EnterSignalArgs args) =>
+        {
+            _mousePosition.X      = args.X;
+            _mousePosition.Y      = args.Y;
+            _mouseIsInDrawingArea = true;
+
+            (sender.Widget as DrawingArea)?.QueueDraw();
+        };
+
+        mouseMove.OnLeave  += (EventControllerMotion sender, EventArgs args) =>
+        {
+            _mouseIsInDrawingArea = false;
+
+            (sender.Widget as DrawingArea)?.QueueDraw();
+        };
+
+        return drawingArea;
+    }
+    //-------------------------------------------------------------------------
+    private void Draw(DrawingArea drawingArea, CairoContext cr, int width, int height)
+    {
+        Debug.Assert(width  == Size);
+        Debug.Assert(height == Size);
+
+        using (cr.Save())
+        {
+            cr.Rectangle(0, 0, width, height);
+            cr.LineWidth = 1d;
+            cr.Stroke();
+        }
+
+        if (_functionImageSurface is null)
+        {
+#if USE_NEW_IMAGE_SURFACE_FOR_DATA
+            _functionImageSurface = new ImageSurface(Format.Argb32, Size, Size);
+#else
+            _functionImageSurface = cr.Target.CreateSimilarImage(Format.Argb32, Size, Size);
+#endif
+            Plotter.CreateFunctionSurface<TurboColorMap>(_functionImageSurface, s_funcData, s_funcMin, s_funcMax);
+        }
+
+        cr.SetSourceSurface(_functionImageSurface, 0, 0);
+        cr.Paint();
```

Comment:
In a cairo mailing list I read that such a "copy" should be done as early as possible in order for the backend to open the doors for optimizations (especially for X11 where the pixbuf doesn't need to be copied back and forth).